Testsuite fails on i586 in test_xvfb.TestXvfb.test_start_fails_with_unknown_kwargs

```
[   84s] 
[   84s] ======================================================================
[   84s] FAIL: test_start_fails_with_unknown_kwargs (test_xvfb.TestXvfb)
[   84s] ----------------------------------------------------------------------
[   84s] Traceback (most recent call last):
[   84s]   File "/home/abuild/rpmbuild/BUILD/xvfbwrapper-0.2.9/test_xvfb.py", line 89, in test_start_fails_with_unknown_kwargs
[   84s]     xvfb.start()
[   84s] AssertionError: RuntimeError not raised
```

[Full build log](https://github.com/cgoldberg/xvfbwrapper/files/2450676/python-xvfbwrapper-build-log-i586.txt)

--- a/xvfbwrapper.py
+++ b/xvfbwrapper.py
@@ -20,7 +20,6 @@ class Xvfb(object):
     # Maximum value to use for a display. 32-bit maxint is the
     # highest Xvfb currently supports
     MAX_DISPLAY = 2147483647
-    SLEEP_TIME_BEFORE_START = 0.1
 
     def __init__(self, width=800, height=680, colordepth=24, tempdir=None,
                  display=None, environ=None, **kwargs):
@@ -69,21 +68,20 @@ class Xvfb(object):
         else:
             self.new_display = self._get_next_unused_display()
         display_var = ':{}'.format(self.new_display)
-        self.xvfb_cmd = ['Xvfb', display_var] + self.extra_xvfb_args
-        with open(os.devnull, 'w') as fnull:
-            self.proc = subprocess.Popen(self.xvfb_cmd,
-                                         stdout=fnull,
-                                         stderr=fnull,
-                                         close_fds=True)
-        # give Xvfb time to start
-        time.sleep(self.__class__.SLEEP_TIME_BEFORE_START)
+        xvfb_cmd = ['Xvfb', display_var] + self.extra_xvfb_args
+        self.proc = subprocess.Popen(xvfb_cmd,
+                                     stdout=subprocess.DEVNULL,
+                                     stderr=subprocess.DEVNULL,
+                                     close_fds=True)
+        while not local_display_exists(self.new_display):
+            time.sleep(1e-3)
         ret_code = self.proc.poll()
         if ret_code is None:
             self._set_display(display_var)
         else:
             self._cleanup_lock_file()
             raise RuntimeError('Xvfb did not start ({0}): {1}'
-                               .format(ret_code, self.xvfb_cmd))
+                               .format(ret_code, xvfb_cmd))
 
     def stop(self):
         try:
@@ -165,3 +163,7 @@ class Xvfb(object):
 
     def _set_display(self, display_var):
         self.environ['DISPLAY'] = display_var
+
+
+def local_display_exists(display):
+    return os.path.exists('/tmp/.X11-unix/X{}'.format(display))

--- a/xvfbwrapper.py
+++ b/xvfbwrapper.py
@@ -22,11 +22,12 @@ class Xvfb(object):
     MAX_DISPLAY = 2147483647
 
     def __init__(self, width=800, height=680, colordepth=24, tempdir=None,
-                 display=None, environ=None, **kwargs):
+                 display=None, environ=None, timeout=5, **kwargs):
         self.width = width
         self.height = height
         self.colordepth = colordepth
         self._tempdir = tempdir or tempfile.gettempdir()
+        self._timeout = timeout
         self.new_display = display
 
         if environ:
@@ -73,8 +74,11 @@ class Xvfb(object):
                                      stdout=subprocess.DEVNULL,
                                      stderr=subprocess.DEVNULL,
                                      close_fds=True)
+        start = time.time()
         while not local_display_exists(self.new_display):
             time.sleep(1e-3)
+            if time.time() - start > self._timeout:
+                raise TimeoutError
         ret_code = self.proc.poll()
         if ret_code is None:
             self._set_display(display_var)
@@ -92,7 +96,7 @@ class Xvfb(object):
             if self.proc is not None:
                 try:
                     self.proc.terminate()
-                    self.proc.wait()
+                    self.proc.wait(self._timeout)
                 except OSError:
                     pass
                 self.proc = None

--- a/xvfbwrapper.py
+++ b/xvfbwrapper.py
@@ -39,6 +39,7 @@ class Xvfb(object):
             msg = 'Can not find Xvfb. Please install it and try again.'
             raise EnvironmentError(msg)
 
+        self.xvfb_cmd = []
         self.extra_xvfb_args = ['-screen', '0', '{}x{}x{}'.format(
                                 self.width, self.height, self.colordepth)]
 
@@ -69,8 +70,8 @@ class Xvfb(object):
         else:
             self.new_display = self._get_next_unused_display()
         display_var = ':{}'.format(self.new_display)
-        xvfb_cmd = ['Xvfb', display_var] + self.extra_xvfb_args
-        self.proc = subprocess.Popen(xvfb_cmd,
+        self.xvfb_cmd = ['Xvfb', display_var] + self.extra_xvfb_args
+        self.proc = subprocess.Popen(self.xvfb_cmd,
                                      stdout=subprocess.DEVNULL,
                                      stderr=subprocess.DEVNULL,
                                      close_fds=True)
@@ -79,14 +80,14 @@ class Xvfb(object):
             time.sleep(1e-3)
             if time.time() - start > self._timeout:
                 self.stop()
-                raise RuntimeError('Xvfb display did not open: {}'.format(xvfb_cmd))
+                raise RuntimeError('Xvfb display did not open: {}'.format(self.xvfb_cmd))
         ret_code = self.proc.poll()
         if ret_code is None:
             self._set_display(display_var)
         else:
             self._cleanup_lock_file()
             raise RuntimeError('Xvfb did not start ({0}): {1}'
-                               .format(ret_code, xvfb_cmd))
+                               .format(ret_code, self.xvfb_cmd))
 
     def stop(self):
         try:

--- a/xvfbwrapper.py
+++ b/xvfbwrapper.py
@@ -78,7 +78,8 @@ class Xvfb(object):
         while not local_display_exists(self.new_display):
             time.sleep(1e-3)
             if time.time() - start > self._timeout:
-                raise TimeoutError
+                self.stop()
+                raise RuntimeError('Xvfb display did not open: {}'.format(xvfb_cmd))
         ret_code = self.proc.poll()
         if ret_code is None:
             self._set_display(display_var)

